pynest/nest/tests/test_stdp_synapse.py

```diff
@@ -0,0 +1,360 @@
+                                       weight_by_nest,
+                                       fname_snip=fname_snip,
+                                       title_snip=self.nest_neuron_model + " (NEST)")
+
+        t_weight_reproduced_independently, weight_reproduced_independently = self.reproduce_weight_drift(
+            pre_spikes, post_spikes,
+            self.init_weight,
+            fname_snip=fname_snip)
+
+        # ``weight_by_nest`` containts only weight values at pre spike times, ``weight_reproduced_independently``
+        # contains the weight at pre *and* post times: check that weights are equal only for pre spike times
+        assert len(weight_by_nest) > 0
+        for idx_pre_spike_nest, t_pre_spike_nest in enumerate(t_weight_by_nest):
+            idx_pre_spike_reproduced_independently = \
+                np.argmin((t_pre_spike_nest - t_weight_reproduced_independently)**2)
+            np.testing.assert_allclose(t_pre_spike_nest,
+                                       t_weight_reproduced_independently[idx_pre_spike_reproduced_independently])
+            np.testing.assert_allclose(weight_by_nest[idx_pre_spike_nest],
+                                       weight_reproduced_independently[idx_pre_spike_reproduced_independently])
+
+    def do_the_nest_simulation(self):
+        """
+        This function is where calls to NEST reside. Returns the generated pre- and post spike sequences and the
+        resulting weight established by STDP.
+        """
+        nest.set_verbosity('M_WARNING')
+        nest.ResetKernel()
+        nest.SetKernelStatus({'resolution': self.resolution})
+
+        neurons = nest.Create(
+            self.nest_neuron_model,
+            2,
+            params=self.neuron_parameters)
+        presynaptic_neuron = neurons[0]
+        postsynaptic_neuron = neurons[1]
+
+        generators = nest.Create(
+            "poisson_generator",
+            2,
+            params=({"rate": self.presynaptic_firing_rate,
+                     "stop": (self.simulation_duration - self.hardcoded_trains_length)},
+                    {"rate": self.postsynaptic_firing_rate,
+                     "stop": (self.simulation_duration - self.hardcoded_trains_length)}))
+        presynaptic_generator = generators[0]
+        postsynaptic_generator = generators[1]
+
+        wr = nest.Create('weight_recorder')
+        nest.CopyModel(self.synapse_model, self.synapse_model + "_rec", {"weight_recorder": wr})
+
+        # While the random sequences, fairly long, would supposedly
+        # reveal small differences in the weight change between NEST
+        # and ours, some low-probability events (say, coinciding
+        # spikes) can well not have occured. To generate and
+        # test every possible combination of pre/post precedence, we
+        # append some hardcoded spike sequences:
+        # pre: 1       5 6 7   9    11 12 13
+        # post:  2 3 4       8 9 10    12
+        (
```

Comment:
The python magic beginning here is rather too tough for my little head. Are you somehow shifting spike times along the time axis? Why is that necessary, and couldn't you operate directly on the numpy arrays instead of the acrobatics here?